### DISCUSSION TITLE
Unable to set MTU size = 9000 bytes for NIC interface in the Ethernet transmitter

### DISCUSSION THREAD

**Discussion by wei-lou** [srsran/srsRAN_Project#376] [General Help] [answered]
Hi there, I compiled the latest 23.10.1 load on ubuntu 22.04 server VM. The NIC was passed through to the VM. However every time i am trying to run the code, it give me such error.

"labuser@srsran23101r:~$ sudo srsRAN_Project/build/apps/gnb/gnb -c srsRAN_Project/configs/gnb_ru_fuji_tdd_n48_40mhz_2x2_23101R.yml
```

The PRACH detector will not meet the performance requirements with the configuration {Format 0, ZCZ 0, SCS 1.25kHz, Rx ports 1}.

--== srsRAN gNB (commit 55c984b55) ==--

Connecting to AMF on 192.168.70.210:38412
Initializing Open Fronthaul Interface sector=0: ul_comp=[BFP,9], dl_comp=[BFP,9], prach_comp=[BFP,9] prach_cp_enabled=false, downlink_broadcast=false.
srsGNB ERROR: Unable to set MTU size = 9000 bytes for NIC interface in the Ethernet transmitter, current MTU = -1 bytes
Aborted (core dumped)


``` 
"
I have another VM running 23.10.1.Test load and I don't have such problem. Please find attached gnb.log. 
[MTU_Check_Failed_gnb.log](https://github.com/srsran/srsRAN_Project/files/13717408/MTU_Check_Failed_gnb.log)

Thanks,

**Comment by andrepuschmann** [comment]
Hey @wei-lou - I don't think you can set the MTU in the VM, at least not for VFs. The code tries that but fails. You could try to comment the section in `ethernet_transmitter_impl.cc` to see if that works. Or there is an option to allow setting the MTU from within the VM.

**Reply by wei-lou** [reply]
Hi @andrepuschmann  I commented out that section of checking, now it gets error on NIC index, seems like i need to comment that section out as well. Not after these will cause any issues later or not.
srsGNB ERROR: Unable to get index for NIC interface in the Ethernet transmitter

BTW, i compiled the code on bare metal server without VM, I am still getting the same error as the first one
"Connecting to AMF on 192.168.70.210:38412
Initializing Open Fronthaul Interface sector=0: ul_comp=[BFP,9], dl_comp=[BFP,9], prach_comp=[BFP,9] prach_cp_enabled=true, downlink_broadcast=false.
srsGNB ERROR: Unable to set MTU size = 9000 bytes for NIC interface in the Ethernet transmitter, current MTU = -1 bytes
Aborted (core dumped)"

**Reply by andrepuschmann** [reply]
This is an environment issue with the VM. But yes, if the first `ioctl` call fails, the second is likely to fail too. 

But on bare metal this should not happen. Do you run the app with sudo? Can you share the output of `ifconfig` for the interface you're using?

**Reply by wei-lou** [reply]
Hi @andrepuschmann, I wasn't sure what I was talking about the bare metal observation. Maybe I got too many VMs and bare metal setups that I was confused with the issue. Actually the bare metal server installation seems OK. The issue was with the VM installation. 

In the previous attempts, I only commented out this section. 
```c++


/*
  // Set requested MTU size.
  if_idx.ifr_mtu = config.mtu_size.value();
  if (::ioctl(socket_fd, SIOCSIFMTU, &if_idx) < 0) {
    // Get the MTU size of the NIC.
    int current_mtu = -1;
    if (::ioctl(socket_fd, SIOCGIFMTU, &if_idx) < 0) {
      logger.warning("Could not check MTU of the NIC interface in the Ethernet transmitter");
    } else {
      current_mtu = if_idx.ifr_mtu;
    }
    report_error(
        "Unable to set MTU size = {} bytes for NIC interface in the Ethernet transmitter, current MTU = {} bytes",
        config.mtu_size,
        current_mtu);
  }
*/
``` 
Would I comment out next section which checking the NIC index work? I haven't tried that since recompile takes time on my VM.

**Reply by andrepuschmann** [reply]
It's unlikely to work if the previous `ioctl` call fails. I assume there is an issue with your network settings in the VM. Can you share your configuration please?

**Reply by wei-lou** [reply]
ens16 is the fronthaul interface (25G). VLAN 201 for m-plane and VLAN 202 for CU-Plane


lab@srsran23101r:~$ ifconfig
```


ens16: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 9000
        inet6 fe80::3efd:feff:feed:d320  prefixlen 64  scopeid 0x20<link>
        ether 3c:fd:fe:ed:d3:20  txqueuelen 1000  (Ethernet)
        RX packets 20864  bytes 5736309 (5.7 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 4519  bytes 589630 (589.6 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

ens18: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.70.208  netmask 255.255.255.0  broadcast 192.168.70.255
        inet6 fe80::44a:faff:fed5:4b0c  prefixlen 64  scopeid 0x20<link>
        ether 06:4a:fa:d5:4b:0c  txqueuelen 1000  (Ethernet)
        RX packets 17713  bytes 9201956 (9.2 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 22126  bytes 3301678 (3.3 MB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

ens16.201: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 9000
        inet 192.168.1.108  netmask 255.255.255.0  broadcast 192.168.1.255
        inet6 fe80::3efd:feff:feed:d320  prefixlen 64  scopeid 0x20<link>
        ether 3c:fd:fe:ed:d3:20  txqueuelen 1000  (Ethernet)
        RX packets 11775  bytes 2463021 (2.4 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 4439  bytes 583798 (583.7 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

ens16.202: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 9000
        inet6 fe80::3efd:feff:feed:d320  prefixlen 64  scopeid 0x20<link>
        ether 3c:fd:fe:ed:d3:20  txqueuelen 1000  (Ethernet)
        RX packets 9089  bytes 2981192 (2.9 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 41  bytes 3006 (3.0 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 174  bytes 71683 (71.6 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 174  bytes 71683 (71.6 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

``` 

The netplan yaml file is as follow:
```

# This is the network config written by 'subiquity'
network:
  ethernets:
    ens16:
      addresses: []
      mtu: 9000
    ens18:
      addresses:
      - 192.168.70.208/24
      nameservers:
        addresses:
        - 192.168.70.100
        search: []
      routes:
      - to: default
        via: 192.168.70.250
  vlans:
    ens16.201:
      id: 201
      link: ens16
      addresses: [192.168.1.108/24]
    ens16.202:
      id: 202
      link: ens16
  version: 2

``` 

lab@srsran23101r:/etc/netplan$  lspci | awk '/[Nn]et/ {print $1}' | xargs -i% lspci -ks %
```

00:10.0 Ethernet controller: Intel Corporation Ethernet Controller XXV710 for 25GbE SFP28 (rev 02)
        Subsystem: Intel Corporation Ethernet Network Adapter XXV710-2
        Kernel driver in use: i40e
        Kernel modules: i40e
00:12.0 Ethernet controller: Red Hat, Inc. Virtio network device
        Subsystem: Red Hat, Inc. Virtio network device
        Kernel driver in use: virtio-pci

``` 
The 25G is the passthrough NIC interface and Virtio network device is the virtual bridge interface for VM management

**Reply by andrepuschmann** [reply]
But this is the bare metal config, not the VM's network config. You said bare metal it's working.

Please also check if you can get the formatting right using markdown. It's hard to read otherwise.

**Reply by wei-lou** [reply]
Hi @andrepuschmann. this is the VM installation. There two interfaces for the VM, ens16 which is 25G passthrough interface and ens18 is the VM virtual interface.

**Reply by andrepuschmann** [reply]
I don't understand why this isn't working if it's full PCI passthrough. Those `ioctl` syscalls are nothing special. We'll see if we can replicate it with Proxmox ourselves.

**Reply by wei-lou** [reply, marked as answer]
Thanks for checking. With 23.5 load, I didn't encounter this issue with VM installation. I skipped 23.10, so I didn't with that load there is issue or not.
